File: JavaScript/42. Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    var n= height.length;
    var sol=0;
    l=new Array(n).fill(0);
    r=new Array(n).fill(0);
    l[0]=height[0];
    r[n-1]=height[n-1];
    for(var i=1; i<n; i++){
        l[i]= Math.max(l[i-1],height[i]);
    }
    for(var i=n-2; i>=0; i--){
        r[i]= Math.max(r[i+1],height[i]);
    }
    for(var i=0; i<n; i++){
        sol+= (Math.min(l[i], r[i]))-height[i];
    }
    return sol;
};

File: Java/42. Trapping Rain Water.java
class Solution {
    public int trap(int[] height) {
        int n= height.length;
        int sol=0;
        int[] l=new int[n];
        int[] r=new int[n];
        l[0]=height[0];
        r[n-1]=height[n-1];
        for(int i=1; i<n; i++){
            l[i]= Math.max(l[i-1],height[i]);
        }
        for(int i=n-2; i>=0; i--){
            r[i]= Math.max(r[i+1],height[i]);
        }
        for(int i=0; i<n; i++){
            sol+= (Math.min(l[i], r[i]))-height[i];
        }
        return sol;
    }
}
